feat: Support authentication via Shared Access Signatures (SAS) for Azure artifacts (Fixes #10297) (#13360)
argoproj/argo-workflows: docs/configure-artifact-repository.md

ways to accomplish this, including the [Azure Portal](https://portal.azure.com) or the
[CLI](https://docs.microsoft.com/en-us/cli/azure/).

There are two ways to allow argo to authenticate it's access to your Azure storage account.

> ### Using Azure access keys

Reviewer: This makes it seem like it's either access key or SAS. However, there's also the `useSDKCreds` option where you don't need a secret set and it will use a `DefaultAzureCredential` which supports a number of methods of authentication. I would call that option out first, since it's probably the most secure and preferred auth method. Then access key and SAS are the other two options.

Author: Yes that makes sense! Let me push a change to the doc!

      endpoint: https://mystorageaccountname.blob.core.windows.net
      container: my-container-name
      blob: path/in/container
      # If a managed identity has been assigned to the machines running the
      # workflow (e.g., https://docs.microsoft.com/en-us/azure/aks/use-managed-identity)
      # then useSDKCreds should be set to true. The accountKeySecret is not required
      # and will not be used in this case.
      useSDKCreds: true  
```

> In addition to managed identities, Argo workflows also support authentication using access keys and SAS tokens.

Reviewer 2: ```suggestion
You can also authenticate with [Access Keys](#using-azure-access-keys) and [Shared Access Signatures](#using-azure-shared-access-signatures].
```

- simplify, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-simple-and-direct-language)
- use active voice, per [k8s style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)
- Access Keys is a proper noun, no?

- also link to the two sections and use the full name for SAS as it has not been introduced yet.

Author: Thanks @agilgur5 ! I'll raise a separate PR for this!

Reviewer 2: Addressed in #13376

    ```bash
    az storage account keys list -n mystorageaccountname --query '[0].value' -otsv
    ```

3. Create a Kubernetes secret to hold the storage account key. For example:

    ```bash
    kubectl create secret generic my-azure-storage-credentials \
      --from-literal "account-access-key=$(az storage account keys list -n mystorageaccountname --query '[0].value' -otsv)"
    ```

4. Configure `azure` artifact as follows in the yaml.

    ```yaml
    artifacts:
      - name: message
        path: /tmp/message
        azure:
          endpoint: https://mystorageaccountname.blob.core.windows.net
          container: my-container-name
          blob: path/in/container
          # accountKeySecret is a secret selector.
          # It references the k8s secret named 'my-azure-storage-credentials'.
          # This secret is expected to have the key 'account-access-key',
          # containing the base64 encoded credentials to the storage account.
          accountKeySecret:
            name: my-azure-storage-credentials
            key: account-access-key     
    ```

### Using Azure Shared Access Signatures (SAS)

> 

Reviewer 2: This needs a version specifier a la #12581 et al

```suggestion

> v3.6 and after

```

Reviewer 2: Addressed in #13376
